xmrstak/backend/amd/autoAdjust.hpp

```diff
@@ -97,34 +95,46 @@ class autoAdjust
 		int i = 0;
 		for(auto& ctx : devVec)
 		{
-			/* 1000 is a magic selected limit, the reason is that more than 2GiB memory
-			 * sowing down the memory performance because of TLB cache misses
-			 */
-			size_t maxThreads = 1000u;
-			if(ctx.name.compare("gfx901") == 0)
-			{
-				/* Increase the number of threads for AMD VEGA gpus.
-				 * Limit the number of threads based on the issue: https://github.com/fireice-uk/xmr-stak/issues/5#issuecomment-339425089
-				 * to avoid out of memory errors
-				 */
-				maxThreads = 2024u;
-			}
-
-			// keep 128MiB memory free (value is randomly chosen)
-			size_t availableMem = ctx.freeMem - (128u * byteToMiB);
-			// 224byte extra memory is used per thread for meta data
+			/*
+			  GCN has a hard architectural limit of 64 CUs max (Invalid for other archs though).
+			  Reserve 1 CU for the system. 
+			*/
+			size_t maxComputeUnitsAvailable = std::max(std::min(ctx.computeUnits, 64) - 1, 1);
+			//Each CU is 64 threads executed in an SIMT fashion:
```

Comment:
A wavefront on a AMD gpu is 64 threads. This not means that you should only use 64 threads per multiprozessor. If you oversubscribe the multiprocessor you can hide memory latencies.

Reply:
Totally agree. 

I just found empirically that cryptonight is the odd man out here in that any form of context swapping seems to add load to the memory bottleneck. I should probably have been clearer in the comment I’m explicitly trying to avoid the context switching, even if nearly free (10 wavefronts can be buffered per SIMD unit in each CU I think...).